pkg/kv/kvserver/replica_raftlog_test.go

```diff
@@ -0,0 +1,303 @@
+	rng := newRNG(seed)
+
+	// Start separate threads for truncations and reads under mu. Use stable seed
+	// for each thread.
+	// NB: stable seeds do not achieve reproducibility here, but we try hard. Note
+	// that at least the append thread below is reproducible.
+	for i := 0; i < 5; i++ {
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "read"},
+			func(ctx context.Context) { rt.readLoop(ctx, fork) }))
+	}
+	{
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "trunc"},
+			func(ctx context.Context) { rt.truncateLoop(ctx, fork) }))
+	}
+	// Run the append loop synchronously, and terminate all other tasks when done.
+	rt.writeLoop(ctx, rng, 1000)
+	cancel()
+}
+
+type replicaLogStorageTest struct {
+	t *testing.T
+
+	stopper *stop.Stopper
+
+	ls *replicaLogStorage
+	mu struct {
+		syncutil.RWMutex
+		committed uint64
+		readable  uint64
+	}
+	raftMu syncutil.Mutex
+
+	term uint64
+	ms   *raft.MemoryStorage
+}
+
+func newReplicaLogStorageTest(t *testing.T) *replicaLogStorageTest {
+	const rangeID = 10
+	rt := &replicaLogStorageTest{
+		t:       t,
+		stopper: stop.NewStopper(),
+		term:    1,
+		ms:      raft.NewMemoryStorage(),
+	}
+
+	st := cluster.MakeTestingClusterSettings()
+	eng := storage.NewDefaultInMemForTesting()
+	sideloaded := logstore.NewDiskSideloadStorage(st, rangeID,
+		eng.GetAuxiliaryDir(), nil /* limiter: unused */, eng)
+
+	rt.ls = &replicaLogStorage{
+		ctx:     context.Background(),
+		cache:   raftentry.NewCache(1 << 5), // stress eviction
+		onSync:  noopSyncCallback{},
+		metrics: newStoreMetrics(metric.TestSampleInterval),
+		ls: &logstore.LogStore{
+			RangeID:     rangeID,
+			Engine:      eng,
+			Sideload:    sideloaded,
+			StateLoader: logstore.NewStateLoader(rangeID),
+			SyncWaiter:  logstore.NewSyncWaiterLoop(),
+			Settings:    st,
+		},
+	}
+	rt.ls.mu.RWMutex = &rt.mu.RWMutex
+	rt.ls.raftMu.Mutex = &rt.raftMu
+	return rt
+}
+
+func (rt *replicaLogStorageTest) start(ctx context.Context) {
+	rt.ls.ls.SyncWaiter.Start(ctx, rt.stopper)
+}
+
+func (rt *replicaLogStorageTest) stop(ctx context.Context) {
+	rt.stopper.Stop(ctx)
+	rt.ls.ls.Engine.Close()
+
+	cache := rt.ls.cache.Metrics()
+	rt.t.Logf("cache loads: %d", cache.Accesses.Count())
+	rt.t.Logf("cache hits: %d", cache.Hits.Count())
+	rt.t.Logf("read bytes: %d", rt.ls.metrics.RaftStorageReadBytes.Count())
+
+	rt.mu.RLock()
+	defer rt.mu.RUnlock()
+	require.Equal(rt.t, rt.ms.LastIndex(), rt.mu.readable)
+	rt.t.Logf("term=%d, log=(%d, %d], committed=%d",
+		rt.term, rt.ms.Compacted(), rt.ms.LastIndex(), rt.mu.committed)
+}
+
+func (rt *replicaLogStorageTest) append(ctx context.Context, app raft.StorageAppend) {
+	// While a write is in flight, raft will never attempt to read the affected
+	// log indices. In reality, there can be multiple writes in flight, and raft
+	// tracks the lowest "protected" index. We don't do it here for simplicity,
```

Comment:
This sentence might be confusing. This is just saying that we don't do multiple in-flight writes. But we still need to "protect" the in-flight write from reads, hence the `readable` var. This is the equivalent of `unstable` - all the writes are underneath it.